src/libopensc/card-piv.c

```diff
@@ -467,208 +795,1885 @@ static struct sc_card_driver piv_drv = {
+	r = piv_general_io(card, 0x87, cs->p1, 0x04, sbuf, (p - sbuf), rbuf, rbuflen);
+	if (r <= 0)
+		goto err;
+
+	rbuflen = r;
+	p = rbuf;
+	
+	body = sc_asn1_find_tag(card->ctx, rbuf, rbuflen, 0x7C, &bodylen);
+	if (body == NULL || bodylen < 20 || rbuf[0] != 0x7C) {
+		sc_log(card->ctx, "SM response data to short");
+		r = SC_ERROR_SM_NO_SESSION_KEYS;
+		goto err;
+	}
+
+	payload = sc_asn1_find_tag(card->ctx, body, bodylen, 0x82, &payloadlen);
+	if (payload == NULL || payloadlen < 1 + cs->Nicclen + cs->AuthCryptogramlen || *body != 0x82) {
+		sc_log(card->ctx, "SM response data to short");
+		r = SC_ERROR_SM_NO_SESSION_KEYS;
+		goto err;
+	}
+
+	/* payload is CBicc (1) || Nicc (16 or 24) || AuthCryptogram (CMAC 16 or 16) ||Cicc (variable) */
+	 p = (u8 *) payload;
+
+	/* Step H4 check CBicc == 0x00 */
+	CBicc = *p++;
+	if  (CBicc != 0x00) { /* CBicc must be  zero */
+		sc_log(card->ctx, "SM card did not accept request");
+		r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+		goto err;
+	}
+
+	memcpy(Nicc, p, cs->Nicclen);
+	p += cs->Nicclen;
+
+	memcpy(AuthCryptogram, p, cs->AuthCryptogramlen);
+	p += cs->AuthCryptogramlen;
+
+	if (p > payload + payloadlen) {
+		sc_log(card->ctx, "SM card CVC is to short");
+		r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+		goto err;
+	}
+	cvclen = len = payloadlen - (p - payload);
+	if (len) {
+		cvcder = p; /* in rbuf */
+	
+		r = piv_decode_cvc(card, &p, &len, &priv->sm_cvc);
+		if (r != SC_SUCCESS) {
+			r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+			goto err;
+		}
+		priv->sm_flags |= PIV_SM_FLAGS_SM_CVC_PRESENT;
+	}
+
+	/* Step H5 Verify Cicc CVC and pubkey */
+	/* Verify Cicc (sm_cvc) is signed by sm_in_cvc or PIV_OBJ_SM_CERT_SIGNER  */
+	/* sm_in_cvc is signed by PIV_OBJ_SM_CERT_SIGNER */
+
+	/* Verify the cert chain is valid. */
+	r = piv_sm_verify_certs(card);
+	if (r < 0) {
+		sc_log(card->ctx, "SM  piv_sm_verify_certs r:%d", r);
+		r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+		goto err;
+	}
+	
+	/* Step H6  need left most 8 bytes of hash of sm_cvc */
+	{
+		u8 hash[SHA384_DIGEST_LENGTH] = {0}; /* 384 is max */
+		const u8* tag;
+		size_t taglen;
+		const u8* tmpder;
+		size_t tmpderlen;
+
+		if ((tag = sc_asn1_find_tag(card->ctx, cvcder, cvclen, 0x7F21, &taglen)) == NULL
+				|| *cvcder != 0x7F || *(cvcder + 1) != 0x21) {
+
+			r = SC_ERROR_INTERNAL;
+			goto err;
+		}
+
+		/* debug choice */
+		tmpder = (1==1)? cvcder:tag;
+		tmpderlen = (1==1)?cvclen:taglen;
+
+		if (EVP_DigestInit(hash_ctx,EVP_sha256()) != 1
+				|| EVP_DigestUpdate(hash_ctx, tmpder, tmpderlen) != 1
+				|| EVP_DigestFinal_ex(hash_ctx, hash, NULL) != 1) {
+			sc_log(card->ctx,"IDsicc hash failed");
+			piv_log_openssl(card->ctx);
+			r = SC_ERROR_INTERNAL;
+			goto err;
+		}
+		memcpy(IDsicc, hash, sizeof(IDsicc)); /* left 8 bytes */
+	}
+
+	/* TODO check Ciss uses same curve as ours will fail to authenticate if different */
+	/* Step H7 get the cards public key Qsicc into OpenSSL Cicc_eckey */
+
+#if OPENSSL_VERSION_NUMBER < 0x30000000L
```

Comment:
What are we doing for OpenSSL 3.0?  sc-ossl-compat.h was introduced when there were simple changes that could be handled by simple inline macros.  OpenSSL 3.0 features would be hard to implement in sc-ossl-compat.h

Code already limits OpenSSL versions to above 1.1 https://github.com/dengert/OpenSC/blob/PIV-4-extensions/src/libopensc/card-piv.c#L62-L65 

I would like to leave that until OpenSSL 3.0 is addressed in the rest of OpenSC.

Reply:
If I'm not mistaken, you have two cases in which you're detecting OpenSSL 3.0+:
1. In case of some ECC key initialization (e.g. 1939, 2124, 2315)
2. In case of using CMAC

If I'm not mistaken, the API you're using in the first case for openssl < 3.0 is not deprecated and could still be used. I don't really see a benefit over choosing the newer style.

In the second case the low level CMAC API was indeed deprecated. However, this is the only use of CMAC in OpenSSL, so you should be free of adding CMAC wrappers to  sc-ossl-compat.h without causing conflicts with #2337

Reply:
Eventually we will have to go to to OpenSSL 3.0+, and easier to add it now while 800-73-4 SM was fresh in mind.

Reply:
The problem with 3.0 is very similar with the 1.1.1 -- they are not backward compatible, if you turn off the deprecated functions. Before that, most of the structs were kept in place and only some functions were missing.

Now if we want to support both 1.1.1 and 3.0 we either need to use the old API with 3.0 deprecated functions (problems in the future building with 3.0) or use the ifdefs (and thoughtfully test on both versions).

The last option would be to implement some middle-layer for common operations with common structures, which is something that is attempted to be done in the sm code that I was rewriting, but it worked only partially and that it would cause even more headache than using the ifdefs (but it might be worth trying). This would make sense if we would support for example some different crypto backend (for example Botan in SoftHSM or libgcrypt in libssh -- I can not recommend neither one), but there is no such need in opensc as far as I know.

As far as I know, initialization of EC keys and EC_KEY structure is deprecated in 3.0 and only accessing them through the EVP_PKEY is "the new correct way". Similar for the CMAC. But if the uses are with similar enough inputs and outputs so they can be wrapped in macros/intermediate functions, I think this is also an option.